pkg/distributor/distributor.go

```diff
@@ -415,12 +539,46 @@ func (d *Distributor) sendSamplesErr(ctx context.Context, ingester ring.Instance
 
 	_, err = c.(logproto.PusherClient).Push(ctx, req)
 	d.ingesterAppends.WithLabelValues(ingester.Addr).Inc()
-	if err != nil {
-		d.ingesterAppendFailures.WithLabelValues(ingester.Addr).Inc()
+
+	if err == nil {
+		return nil
+	}
+
+	d.ingesterAppendFailures.WithLabelValues(ingester.Addr).Inc()
+
+	if d.cfg.ShardStreams.Mode == NeverShardMode {
+		return err
+	}
+
+	// TODO: look for a better way to recognize a per-stream rate limit.
+	if errMsg := err.Error(); strings.Contains(errMsg, "Per stream rate limit exceeded (limit") {
+		d.streamSharding.Inc()
+		shardLimitedStream(errMsg, d.streamSharder)
 	}
 	return err
 }
 
+// shardLimitedStream shards limited stream based on the given per-stream rate limited message.
+func shardLimitedStream(perStreamErrorMsg string, sharder StreamSharder) {
+	re := regexp.MustCompile("\\{(.*?)\\}")
```

Comment:
Can you move this outside of the loop? It gets compile here every time.

Reply:
Done. I added it as an attribute in the Distributor struct that is reused here; WDYT?